diff --git a/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu b/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu
index 5c9dfd03..97ea58b1 100644
--- a/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu
+++ b/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu
@@ -115,6 +115,7 @@ int ggml_cuda_get_device() {
 
 void ggml_cuda_reset_device(int device) {
     ggml_cuda_set_device(device);
+    CUDA_CHECK(cudaDeviceSynchronize());
     CUDA_CHECK(cudaDeviceReset());
 }
 
