diff --git a/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu b/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu
index 5c9dfd03..b364b4e9 100644
--- a/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu
+++ b/ml/backend/ggml/ggml/src/ggml-cuda/ggml-cuda.cu
@@ -97,14 +97,19 @@ void ggml_cuda_error(const char * stmt, const char * func, const char * file, in
 // this is faster on Windows
 // probably because the Windows CUDA libraries forget to make this check before invoking the drivers
 void ggml_cuda_set_device(int device) {
-    int current_device;
+    thread_local int tls_inited_device = -1;
+
+    int current_device = -1;
     CUDA_CHECK(cudaGetDevice(&current_device));
 
-    if (device == current_device) {
-        return;
+    if (device != current_device) {
+        CUDA_CHECK(cudaSetDevice(device));
     }
 
-    CUDA_CHECK(cudaSetDevice(device));
+    if (tls_inited_device != device) {
+        CUDA_CHECK(cudaFree(0));
+        tls_inited_device = device;
+    }
 }
 
 int ggml_cuda_get_device() {
@@ -115,7 +120,7 @@ int ggml_cuda_get_device() {
 
 void ggml_cuda_reset_device(int device) {
     ggml_cuda_set_device(device);
-    CUDA_CHECK(cudaDeviceReset());
+    // CUDA_CHECK(cudaDeviceReset());
 }
 
 static cudaError_t ggml_cuda_device_malloc(void ** ptr, size_t size, int device) {
