diff --git a/ml/backend/ggml/ggml.go b/ml/backend/ggml/ggml.go
index ebcc1d86..d1f5f9c6 100644
--- a/ml/backend/ggml/ggml.go
+++ b/ml/backend/ggml/ggml.go
@@ -625,6 +625,10 @@ func (b *Backend) Load(ctx context.Context, progress func(float32)) error {
 		})
 	}
 
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
 	// Cleanup any backend state from devices that we didn't end up using
 nextDevice:
 	for _, d := range append(gpus, append(accels, cpus...)...) {
@@ -637,10 +641,6 @@ nextDevice:
 		C.ggml_backend_dev_reset(d)
 	}
 
-	if err := g.Wait(); err != nil {
-		return err
-	}
-
 	return nil
 }
 
