#https://bugs.gentoo.org/690828
--- fossil-2.10/autosetup/local.tcl.orig	2020-02-25 22:16:02.166000000 -0600
+++ fossil-2.10/autosetup/local.tcl	2020-02-25 22:19:32.389000000 -0600
@@ -12,7 +12,7 @@
 			return [parse-tclconfig-sh-file $p/tclConfig.sh]
 		}
 		# Some systems allow for multiple versions
-		foreach libpath {lib/tcl8.6 lib/tcl8.5 lib/tcl8.4 lib/tcl tcl lib}  {
+		foreach libpath {lib/tcl8.6 lib/tcl8.5 lib/tcl8.4 lib/tcl tcl lib lib64}  {
 			if {[file exists $p/$libpath/tclConfig.sh]} {
 				return [parse-tclconfig-sh-file $p/$libpath/tclConfig.sh]
 			}
